src/VkBootstrap.cpp

```diff
@@ -1086,89 +1091,92 @@ PhysicalDeviceSelector::Suitable PhysicalDeviceSelector::is_device_suitable(Phys
-	if (criteria.require_separate_compute_queue && !separate_compute) return Suitable::no;
-	if (criteria.require_separate_transfer_queue && !separate_transfer) return Suitable::no;
+	if (criteria.require_dedicated_compute_queue && !dedicated_compute)
+		return PhysicalDevice::Suitable::no;
+	if (criteria.require_dedicated_transfer_queue && !dedicated_transfer)
+		return PhysicalDevice::Suitable::no;
+	if (criteria.require_separate_compute_queue && !separate_compute)
+		return PhysicalDevice::Suitable::no;
+	if (criteria.require_separate_transfer_queue && !separate_transfer)
+		return PhysicalDevice::Suitable::no;
 	if (criteria.require_present && !present_queue && !criteria.defer_surface_initialization)
-		return Suitable::no;
+		return PhysicalDevice::Suitable::no;
 
 	auto required_extensions_supported =
-	    detail::check_device_extension_support(pd.phys_device, criteria.required_extensions);
+	    detail::check_device_extension_support(pd.physical_device, criteria.required_extensions);
 	if (required_extensions_supported.size() != criteria.required_extensions.size())
-		return Suitable::no;
+		return PhysicalDevice::Suitable::no;
 
 	auto desired_extensions_supported =
-	    detail::check_device_extension_support(pd.phys_device, criteria.desired_extensions);
+	    detail::check_device_extension_support(pd.physical_device, criteria.desired_extensions);
 	if (desired_extensions_supported.size() != criteria.desired_extensions.size())
-		suitable = Suitable::partial;
+		suitable = PhysicalDevice::Suitable::partial;
 
-	bool swapChainAdequate = false;
-	if (criteria.defer_surface_initialization) {
-		swapChainAdequate = true;
-	} else if (!instance_info.headless) {
+	if (!criteria.defer_surface_initialization && criteria.require_present) {
 		std::vector<VkSurfaceFormatKHR> formats;
 		std::vector<VkPresentModeKHR> present_modes;
 
 		auto formats_ret = detail::get_vector<VkSurfaceFormatKHR>(formats,
 		    detail::vulkan_functions().fp_vkGetPhysicalDeviceSurfaceFormatsKHR,
-		    pd.phys_device,
+		    pd.physical_device,
 		    instance_info.surface);
 		auto present_modes_ret = detail::get_vector<VkPresentModeKHR>(present_modes,
 		    detail::vulkan_functions().fp_vkGetPhysicalDeviceSurfacePresentModesKHR,
-		    pd.phys_device,
+		    pd.physical_device,
 		    instance_info.surface);
 
-		if (formats_ret == VK_SUCCESS && present_modes_ret == VK_SUCCESS) {
-			swapChainAdequate = !formats.empty() && !present_modes.empty();
+		if (formats_ret != VK_SUCCESS || present_modes_ret != VK_SUCCESS || formats.empty() ||
+		    present_modes.empty()) {
+			return PhysicalDevice::Suitable::no;
 		}
 	}
-	if (criteria.require_present && !swapChainAdequate) return Suitable::no;
 
-	if (pd.device_properties.deviceType != static_cast<VkPhysicalDeviceType>(criteria.preferred_type)) {
+	if (pd.properties.deviceType != static_cast<VkPhysicalDeviceType>(criteria.preferred_type)) {
```

Comment:
While you're here, would it be possible to touch this up such that a user can optionally (I think the current default is good) null this out and say they don't have a preferred type. allow_any_type doesn't quite do this since it will return a partially suitable value for a mismatch.

Reply:
Sure, and no boundaries are being overstepped!
I was hoping you'd review, since you had a good idea of what you *wanted* out of this feature.
I chose to re-do the work myself since its a bit easier to manage + I can fix a half dozen other things I don't like while I work up this feature.

Reply:
I looked at the logic, and think its best to make `allow_any_type` disable the check entirely instead, instead of downgrade to partial. Frankly the whole partial stuff is getting annoying and I have half a mind to rip it out.
```c
//old
if (pd.properties.deviceType != static_cast<VkPhysicalDeviceType>(criteria.preferred_type)) {
	if (criteria.allow_any_type)
		suitable = PhysicalDevice::Suitable::partial;
	else
		return PhysicalDevice::Suitable::no;
}
//new
if (!criteria.allow_any_type &&
	pd.properties.deviceType != static_cast<VkPhysicalDeviceType>(criteria.preferred_type)) {
	suitable = PhysicalDevice::Suitable::partial;
}
```